### PR TITLE
Prevent takeoff in fixed-wing mode for VTOL vehicles

### PR DESCRIPTION
**Pull request by deonblaauw**
Interlock added to prevent takeoff in fixed-wing mode when the vehicle type is VTOL. This is a safety feature and is especially handy for tailsitters.

### CODE REVIEW

File: src/modules/commander/commander.cpp [hunk 1]
```diff
@@ -855,13 +855,22 @@ bool handle_command(struct vehicle_status_s *status_local, const struct safety_s
...
+
 					// Refuse to arm if preflight checks have failed
 					if ((!status_local->hil_state) != vehicle_status_s::HIL_STATE_ON && !status_flags.condition_system_sensors_initialized) {
 						mavlink_log_critical(&mavlink_log_pub, "Arming DENIED. Preflight checks have failed.");
 						cmd_result = vehicle_command_s::VEHICLE_CMD_RESULT_DENIED;
 						break;
 					}
 
+					// Make sure we do not takeoff as a fixed wing when the vehicle type is VTOL
+					if ( status_local->is_vtol && !status_local->is_rotary_wing )
```
Comment (dagar) [inline src/modules/commander/commander.cpp]: Nitpicks
 - no padding in condition
 - brace on same line as if

File: src/modules/commander/state_machine_helper.cpp [hunk 2]
```diff
@@ -1129,6 +1129,14 @@ int preflight_check(struct vehicle_status_s *status, orb_advert_t *mavlink_log_p
 		}
 	}
 
+	if ( status->is_vtol && !status->is_rotary_wing ) {
```
Comment (dagar) [inline src/modules/commander/state_machine_helper.cpp]: Nitpick
 - no padding in condition

**Comment by dagar** [comment]
What about preflight checks for fixed wing on the ground? Arm only as multicopter and transition after?

I don't know if it even works, but what about IN_AIR_RESTORE arming?

It seems like another check we'd want is if you're on the ground and enter mission mode as FW.

File: src/modules/commander/commander.cpp (hunk 1 above)
Comment (dagar) [inline src/modules/commander/commander.cpp]: Isn't this redundant? Won't the same code in preflight_check() catch this?

Comment (deonblaauw) [inline src/modules/commander/commander.cpp]: The preflight_check() code does not run during HITL. We are doing quite a lot of development in HITL, and would really like to have this check in place (makes it more representative of the real-world situation)

**Comment by deonblaauw** [comment]
Updated the code style. In general, preflight_check() does not run in HITL mode, and we really would like to have this interlock in place while testing in our HITL integration lab.

**Comment by dagar** [comment]
There's a submodule change committed I assume isn't intentional?

Wouldn't this also block disarming as FW via command?

Have you found this to mainly be a problem arming straight into a mission, or other modes as well? I'm just wondering if it makes more sense as a check when starting a mission, rather than arming. 

@AndreasAntener any comments?

**Comment by AndreasAntener** [comment]
It should not be a hard-coded interlock. I think it could be the default setting but we need to be able to disable it at least. Where to put the check is another question, I don't have a great idea right now.

**Comment by deonblaauw** [comment]
@dagar The submodule commit was unintentional, no idea how that happened to be honest (apologies). We could have the check on entering mission mode, however that still does not protect against arming and going into "take-off" mode as a fixed wing, which is why we thought blocking on "arming" probably makes more sense. That being said, we are driving towards a fully autonomous solution so we are comfortable with the block on entering "mission" if you guys prefer (we're easy ;) )

**Comment by dagar** [comment]
No problem, on the submodule. They're a frequent source of frustration.

We're just trying to think through the use case for all users. In general I think forcing VTOLs to initially arm as multicopters is acceptable if...
    - it can be disabled
    - doesn't change disarming or in air restart
    - only if landed (depending on how this interacts with in air restore)
    - make the message more appropriate. Arm is not synonymous with takeoff.

After all that it seems like you'd still want an additional check at the beginning of a mission. Perhaps only if that mission actually contains vtol cmds (takeoff, transition, land).

**Comment by LorenzMeier** [comment]
@deonblaauw Note that this will not be merged until the review status is ok.

**Comment by LorenzMeier** [comment]
Please rebase, thanks!

**Comment by deonblaauw** [comment]
Rebased. Have not made any other updates yet, will get to it as soon as I can.

**Comment by sanderux** [comment]
Isnt this already handled by VT_NAV_FORCE_VT?

**Comment by dagar** [comment]
@sanderux only for missions, this would be for any mode.

**Comment by deonblaauw** [comment]
Sorry for not looking into this yet, we are still doing some in-house testing, planning to circle back to this feature in the new year

**Pull request closed without merging**